# Week 2 Checkpoint: First Data Profile
**BAN 6003 Data Management and Analytics Integration**

Complete this notebook using the `employee_data.csv` file in the `data/` folder.

This is the independent checkpoint for Week 2. Use the functions practiced in the guided lab. You are not expected to clean the data yet. Your goal is to profile the data and report what you notice.


## Instructions

1. Run each cell from top to bottom.
2. Fill in all cells marked `# Your code here`.
3. Write short answers in the markdown response areas.
4. Save, commit, and push your notebook to GitHub.

You may use the Week 2 guided lab as a reference.


## Task 0: Import Pandas and Load the Data
Run the cell below.


In [5]:
import pandas as pd

df = pd.read_csv("../data/employee_data.csv")
df.head()


,employee_id,first_name,last_name,hire_date,department,salary,age,full_time
0,101,John,Smith,4/15/22,Marketing,85000,34.0,True
1,102,Jane,Doe,8/20/21,Sales,92000,45.0,True
2,103,Peter,Jones,1/2/22,HR,71000,29.0,True
3,104,Mary,Johnson,11/10/23,Engineering,115000,38.0,True
4,105,Mike,Brown,5/30/20,sales,89000,NaN,False


## Task 1: Basic Structure
Show the basic structure of the DataFrame.

Your code should show:
- the shape of the DataFrame;
- the column names;
- the first five rows;
- the last five rows.


In [9]:
# The shape of the DataFrame
print("DataFrame Shape Attribute")
print(df.shape)
# We can also say:
num_rows, num_columns = df.shape
print(f"This DataFrame has {num_columns} columns, or series, and {num_rows} rows")

#Column (series) Names
print("DataFrame Column Names")
print(df.columns)

#Showing only the first five rows uses the "head" method:
print("DataFrame Head Method")
print(df.head())

#To only use the last five rows, we use the "tail" method:
print("DataFrame Tail Method")
print(df.tail())

#I know the commands are mirror images of the guided lab, but I promise that I am typing them out for practice and not just copy/pasting them!

DataFrame Shape Attribute
(13, 8)
This DataFrame has 8 columns, or series, and 13 rows
DataFrame Column Names
Index(['employee_id', 'first_name', 'last_name', 'hire_date', 'department',
       'salary', 'age', 'full_time'],
      dtype='str')
DataFrame Head Method
  employee_id first_name last_name hire_date   department  salary   age  \
0         101       John     Smith   4/15/22    Marketing   85000  34.0   
1         102       Jane       Doe   8/20/21        Sales   92000  45.0   
2         103      Peter     Jones    1/2/22           HR   71000  29.0   
3         104       Mary   Johnson  11/10/23  Engineering  115000  38.0   
4         105       Mike     Brown   5/30/20        sales   89000   NaN   

  full_time  
0      True  
1      True  
2      True  
3      True  
4     False  
DataFrame Tail Method
   employee_id first_name last_name hire_date   department    salary    age  \
8          102      Emily     White   1/15/24        Sales     91000   31.0   
9          110      

### Task 1 Written Response
Write 1–2 sentences describing the size and general structure of the dataset.

**Your response:**  
The dataset is 13x8, meaning it has 13 rows and 8 series or columns. There are multiple issues with the data consistency that should be corrected before performing any analysis.


## Task 2: Data Types and Non-Missing Counts
Use Pandas to inspect data types and non-missing counts.

Your code should use:
- `info()`
- `dtypes`


In [12]:
# DataFrame info
print(df.info)

#Data types in the DataFram
print(df.dtypes)


<bound method DataFrame.info of    employee_id first_name last_name    hire_date   department    salary  \
0          101       John     Smith      4/15/22    Marketing     85000   
1          102       Jane       Doe      8/20/21        Sales     92000   
2          103      Peter     Jones       1/2/22           HR     71000   
3          104       Mary   Johnson     11/10/23  Engineering    115000   
4          105       Mike     Brown      5/30/20        sales     89000   
5          106      Susan     Davis          NaN        Sales     95000   
6          107      Chris    Wilson  2023-FEB-21    Marketing       NaN   
7          108      Laura    Taylor       9/1/21  Engineering    120000   
8          102      Emily     White      1/15/24        Sales     91000   
9          110      David    Miller      7/22/19           HR     68000   
10         111      Sarah     Moore      3/12/23    Marketing  $91,000    
11       error      Kevin     Clark       2/5/24        Sales     88

### Task 2 Written Response
Which columns, if any, have data types that might need to be investigated later? For example, is a number or date stored as text/object?

**Your response:**  
The "Employee ID", "Hire Date" and "Salary" columns have string data types and should be looked at further as they should only be numerical. 


## Task 3: Missing Values
Check missing values for all columns.

Your code should use `isna().sum()`.

Optional: also calculate missing percentages.


In [15]:
# Checking for missing values using "isna" and "sum" methods
print(df.isna().sum())

# Calculating a percentage
missing_summary = pd.DataFrame({
    "missing_count": df.isna().sum(),
    "missing_percent": df.isna().mean() * 100
})

print(missing_summary)

employee_id    0
first_name     0
last_name      0
hire_date      1
department     0
salary         1
age            1
full_time      1
dtype: int64
             missing_count  missing_percent
employee_id              0         0.000000
first_name               0         0.000000
last_name                0         0.000000
hire_date                1         7.692308
department               0         0.000000
salary                   1         7.692308
age                      1         7.692308
full_time                1         7.692308


### Task 3 Written Response
Which columns have missing values? Which missing values seem most important to investigate later?

**Your response:**  
The "hire_date", "salary", "age", and "full_time" columns all have missing values. While it is important to have consistent data throughout the dataframe for accurate analysis, the "salary", "full_time" and "hire_date" columns could cause one to run into more issues during analysis as they could have financial impacts as well as impacts to an employee's benefits.

## Task 4: Numerical Summary
Use `.describe()` to summarize numerical columns.

Then choose one numerical column and calculate at least two individual summary statistics such as mean, median, min, max, or standard deviation.


In [22]:
# Using the "describe" function
print(df.describe)

#Overacheivers will calculate all of the summary statistics, but we will change the order so that we know we did not just copy and paste them from the lab
print("Standard Deviation:", df["age"].std())
print("Maximum Age:", df["age"].max())
print("Minimum Age:", df["age"].min())
print("Median Age:", df["age"].median())
print("Average Age", df["age"].mean())



<bound method NDFrame.describe of    employee_id first_name last_name    hire_date   department    salary  \
0          101       John     Smith      4/15/22    Marketing     85000   
1          102       Jane       Doe      8/20/21        Sales     92000   
2          103      Peter     Jones       1/2/22           HR     71000   
3          104       Mary   Johnson     11/10/23  Engineering    115000   
4          105       Mike     Brown      5/30/20        sales     89000   
5          106      Susan     Davis          NaN        Sales     95000   
6          107      Chris    Wilson  2023-FEB-21    Marketing       NaN   
7          108      Laura    Taylor       9/1/21  Engineering    120000   
8          102      Emily     White      1/15/24        Sales     91000   
9          110      David    Miller      7/22/19           HR     68000   
10         111      Sarah     Moore      3/12/23    Marketing  $91,000    
11       error      Kevin     Clark       2/5/24        Sales     

### Task 4 Written Response
What do you notice from the numerical summaries? Mention at least one range, unusual value, or pattern.

**Your response:**  
As previously discussed, the "age" column has one entry of 999 which is skewing the rest of the data, throwing off the maximum and average ages.

## Task 5: Categorical Summary
Use `.value_counts()` to summarize at least one categorical column.

Start with `department` if it exists in the dataset. Then choose one additional categorical column if available.


In [28]:
# Using "value counts" to see how many times a value shows up in a categorical category
print(df["department"].value_counts())

print(df["full_time"].value_counts())


department
Sales          4
Marketing      3
Engineering    3
HR             2
sales          1
Name: count, dtype: int64
full_time
True     11
False     1
Name: count, dtype: int64


### Task 5 Written Response
What are the most common categories? Do you notice anything that might need standardization or further investigation later?

**Your response:**  
Sales, Marketing, and Engineering are the most common series for "Departments", and "Sales" has one entry that is not uniform with the rest where the "s" is not capitalized. For the "full time" column, full time was by far the most common category.

## Task 6: Duplicate Checks
Check for duplicate records and duplicate employee IDs.

Your code should check:
- fully duplicated rows;
- duplicate `employee_id` values;
- the most common `employee_id` values using `value_counts()`.


In [34]:
# Count of fully duplicated rows
full_duplicate_rows = df.duplicated().sum()
print(f"Number of fully duplicated rows: {full_duplicate_rows}")

#Count of duplicated employee id's
duplicated_employee_ids = df["employee_id"].duplicated().sum()
print(f"Number of employee ID duplicates: {duplicated_employee_ids}")

#Most Common Employee ID's (we shouldn't have any, but we do!)
print(df["employee_id"].value_counts().head(1))

Number of fully duplicated rows: 0
Number of employee ID duplicates: 1
employee_id
102    2
Name: count, dtype: int64


### Task 6 Written Response
Are there duplicate employee IDs? If yes, what might that mean if `employee_id` is supposed to uniquely identify each employee?

**Your response:**  
I used "head(1)" as we knew from previous examples with this dataframe that there was only one instance of a duplicate ID number. The number "102" has one duplicate and would be a major problem as the "employee ID" column is supposed to be a unique identifier, meaning that each employee ID should only be assigned to one employee or it could cause pandamonium and bad data for analysis.

## Task 7: Overall Data Quality Observations
Write 3–5 bullet points summarizing the condition of the dataset.

Focus on what you found from profiling. Do **not** clean the data yet.

Examples:
- The dataset has ___ rows and ___ columns.
- The ___ column has missing values.
- The ___ column may have an incorrect data type.
- The ___ column has possible duplicate IDs.
- The ___ column has a range from ___ to ___.


**Your observations:**

1. The dataframe has 13 rows and 8 columns, or series.
2. The "Salary" category is not formatted correctly and could cause signficant financial issues during analysis.
3. Multiple columns have missing information that could be easily collected for a dataset this size. The missing information could be problematic during analysis and could skew results.
4. Multiple columns have data formatted incorrectly or is not uniformly formatted whithin the column. 
5. The age column is greatly skewed due to what can only be described as an incorrect entry. 


## Submission Checklist
Before submitting, make sure:

- [ ] All code cells run without errors.
- [ ] You completed all tasks.
- [ ] You wrote short responses in the markdown cells.
- [ ] You saved the notebook.
- [ ] You committed and pushed your work to GitHub.

Use the terminal:

```bash
git add .
git commit -m "Completed Week 2 data profile checkpoint"
git push
```
